### PR TITLE
DEP: Enforce deprecation of date converters for csv

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #35741 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Review by WillAyd — APPROVED** [review]
nice pr

File: pandas/tests/io/parser/test_parse_dates.py [hunk 1]
```diff
@@ -1512,18 +1476,15 @@ def test_generic(all_parsers):
...
-            parse_dates={"ym": [0, 1]},
-            date_parser=lambda y, m: date(year=int(y), month=int(m), day=1),
-        )
-    expected = DataFrame(
-        [[date(2001, 1, 1), 10, 10.0], [date(2001, 2, 1), 1, 11.0]],
-        columns=["ym", "day", "a"],
+    def parse_function(yy, mm):
+        return [date(year=int(y), month=int(m), day=1) for y, m in zip(yy, mm)]
+
+    parser.read_csv(
```
Comment (mroeschke) [inline pandas/tests/io/parser/test_parse_dates.py:1482]: Is it not worth asserting the result of this call anymore?

Comment (phofl) [reply pandas/tests/io/parser/test_parse_dates.py]: No, this was accidental, thx for noticing

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**